### PR TITLE
Command-line redaction + run start and end timestamps

### PR DESCRIPTION
**Pull request by michaelcfanning**

### CODE REVIEW

**Comment by michaelcfanning** [comment]
@lgolding @rtaket I guess I neglected to explicitly solicit review on this :)

File: src/Sarif.Driver/Sdk/AnalyzeCommandBase.cs [hunk 1]
```diff
@@ -178,7 +178,8 @@ private void InitializeOutputFile(TOptions analyzeOptions, TContext context, Has
                                 analyzeOptions.Verbose,
                                 targets,
                                 analyzeOptions.ComputeTargetsHash,
-                                Prerelease)),
+                                Prerelease,
+                                null)),
```
Comment (lgolding) [inline src/Sarif.Driver/Sdk/AnalyzeCommandBase.cs]: `invocationInfoTokensToRedact: null` (I really like the convention of using a named parameter whenever the value doesn't communicate the meaning, such as for literal values.)

File: src/Sarif.Driver/Sdk/SarifLogger.cs [hunk 2]
```diff
@@ -32,36 +33,102 @@ public static ToolInfo CreateDefaultToolInfo(string prereleaseInfo = null)
             return toolInfo;
         }
 
+        private static RunInfo CreateRunInfo(
```
Comment (lgolding) [inline src/Sarif.Driver/Sdk/SarifLogger.cs]: Why place a private method above the public API?

File: src/Sarif.Driver/Sdk/SarifLogger.cs [hunk 3]
```diff
@@ -32,36 +33,102 @@ public static ToolInfo CreateDefaultToolInfo(string prereleaseInfo = null)
...
+                foreach (string target in analysisTargets)
+                {
+                    var fileReference = new FileReference()
+                    {
+                        Uri = target.CreateUriForJsonSerialization(),
+                    };
+
+                    if (computeTargetsHash)
+                    {
+                        string sha256Hash = HashUtilities.ComputeSha256Hash(target) ?? "[could not compute file hash]";
```
Comment (lgolding) [inline src/Sarif.Driver/Sdk/SarifLogger.cs]: It seems to me that there are two options:

1. We care that we couldn't compute the file hash. Then either:
    1. Throw an exception, or
    2. Synthesize a result object to record the problem and add it to the Results list.
2. We don't care that we couldn't compute the file hash. Don't include the hash to the log file.

Replacing the hash with a message like this feels wrong to me. At some point a tool is going to come along, read the log file, compute the hash of the target, compare it to this string, see that it doesn't match, and complain.

File: src/Sarif.Driver/Sdk/SarifLogger.cs [hunk 4]
```diff
@@ -32,36 +33,102 @@ public static ToolInfo CreateDefaultToolInfo(string prereleaseInfo = null)
...
+        {
+            Verbose = verbose;
+
+            _textWriter = writer;
+            _jsonTextWriter = new JsonTextWriter(_textWriter);
+
+            // for debugging it is nice to have the following line added.
+            _jsonTextWriter.Formatting = Newtonsoft.Json.Formatting.Indented;
+            _runInfo = runInfo;
+            _issueLogJsonWriter.WriteToolInfo(toolInfo);
```
Comment (rtaket) [inline src/Sarif.Driver/Sdk/SarifLogger.cs]: Where does the _issueLogJsonWriter get created?

File: src/Sarif/Writers/ResultLogJsonWriter.cs [hunk 5]
```diff
@@ -45,47 +68,50 @@ public ResultLogJsonWriter(JsonWriter jsonWriter)
...
         {
-            if (ruleDescriptors == null) { return; }
+            if (ruleDescriptors == null)
+            {
+                throw new ArgumentNullException("ruleDescriptors");
+            }
+
+            EnsureInitialized();
+            EnsureNoInProgressSerialization();
+            EnsureStateNotAlreadySet(Conditions.Disposed | Conditions.RuleInfoWritten);
```
Comment (rtaket) [inline src/Sarif/Writers/ResultLogJsonWriter.cs]: We check for Conditions.RuleInfoWritten here, but I don't see where we set it.

File: src/Sarif/Writers/ResultLogJsonWriter.cs [hunk 6]
```diff
@@ -124,38 +161,50 @@ public void WriteResult(Result result)
...
+                OpenResults();
             }
 
-            Debug.Assert(_writeState == State.WritingResults);
             _serializer.Serialize(_jsonWriter, result, typeof(Result));
         }
 
         public void CloseResults()
         {
+            EnsureStateNotAlreadySet(Conditions.Disposed);
```
Comment (rtaket) [inline src/Sarif/Writers/ResultLogJsonWriter.cs]: Is this is supposed to be checking for Conditions.ResultsClosed?

**Comment by rtaket** [comment]
Are there other places we need to redact info? e.g. Result message text or exception text written to the log?

**Comment by michaelcfanning** [comment]
Could be. I'm fine with deferring this for now as I see a clear distinction between the data that is used to configure the tool and the data which is expected to be produced as output. Let's tackle the latter separately. The input issue is required to solve an immediate need (for a DB analyzer that would prefer not to capture user/pwd details in logs).

File: src/Sarif.Driver/Sdk/SarifLogger.cs (hunk 4 above)
Comment (lgolding) [inline src/Sarif.Driver/Sdk/SarifLogger.cs]: It looks like it's created in the _other_ overload of the ctor (see below, Line 144), but not in this one. This should be DRYd out -- the ctor with the fewer arguments should call the one with the more arguments, or they should both call a common `Init` method.

File: src/Sarif.Driver/Sdk/SarifLogger.cs [hunk 7]
```diff
@@ -157,12 +193,13 @@ public void LogMessage(bool verbose, string message)
 
         public void AnalysisStarted()
         {
-            // TODO emit a message with timestamp
+            _issueLogJsonWriter.OpenResults();
+            _runInfo.RunStartTime = DateTime.UtcNow;
```
Comment (lgolding) [inline src/Sarif.Driver/Sdk/SarifLogger.cs]: Null-check on `_runInfo`.

File: src/Sarif.UnitTests/Writers/ResultLogJsonWriterTests.cs [hunk 8]
```diff
@@ -34,7 +34,8 @@ private static string GetJson(Action<ResultLogJsonWriter> testContent)
         [TestMethod]
         public void ResultLogJsonWriter_DefaultIsEmpty()
         {
-            Assert.AreEqual(String.Empty, GetJson(delegate { }));
+            string expected = @"{""version"":""1.0.0-beta.1"",""runLogs"":[{}]}";
```
Comment (lgolding) [inline src/Sarif.UnitTests/Writers/ResultLogJsonWriterTests.cs]: Is it "beta.1"? I thought we were now "beta.2" after the big-breaking-change.

File: src/Sarif.UnitTests/Writers/ResultLogJsonWriterTests.cs [hunk 9]
```diff
@@ -86,13 +93,14 @@ public void ResultLogJsonWriter_RequiresNonNullIssue()
...
-                uut.WriteToolAndRunInfo(s_defaultToolInfo, s_defaultRunInfo);
+                uut.WriteToolInfo(s_defaultToolInfo);
+                uut.WriteRunInfo(s_defaultRunInfo);
                 uut.WriteResult(null);
             });
         }
 
         [TestMethod]
-        [ExpectedException(typeof(ObjectDisposedException))]
+        [ExpectedException(typeof(InvalidOperationException))]
```
Comment (lgolding) [inline src/Sarif.UnitTests/Writers/ResultLogJsonWriterTests.cs]: Why this change?

File: src/Sarif.UnitTests/Writers/ResultLogJsonWriterTests.cs [hunk 10]
```diff
@@ -100,20 +108,21 @@ public void ResultLogJsonWriter_CannotWriteToolInfoToDisposedWriter()
...
         public void ResultLogJsonWriter_CannotWriteIssuesToDisposedWriter()
         {
             using (var str = new StringWriter())
             using (var json = new JsonTextWriter(str))
             using (var uut = new ResultLogJsonWriter(json))
             {
-                uut.WriteToolAndRunInfo(s_defaultToolInfo, s_defaultRunInfo);
-                uut.Dispose();
+                uut.WriteToolInfo(s_defaultToolInfo);
+                uut.WriteRunInfo(s_defaultRunInfo); uut.Dispose();
```
Comment (lgolding) [inline src/Sarif.UnitTests/Writers/ResultLogJsonWriterTests.cs]: Line break before `uut.Dispose()`.

File: src/Sarif.UnitTests/Writers/ResultLogObjectWriterTests.cs [hunk 11]
```diff
@@ -70,8 +81,8 @@ public void ResultLogObjectWriter_RequiresNullRunInfoIsOk()
         public void ResultLogObjectWriter_RequiresNonNullIssue()
         {
             var uut = new ResultLogObjectWriter();
-            uut.WriteToolAndRunInfo(s_defaultToolInfo, s_defaultRunInfo);
-            uut.WriteResult(null);
+            uut.WriteToolInfo(s_defaultToolInfo);
+            uut.WriteRunInfo(s_defaultRunInfo); uut.WriteResult(null);
```
Comment (lgolding) [inline src/Sarif.UnitTests/Writers/ResultLogObjectWriterTests.cs]: Line break before `uut.WriteResult`.

File: src/Sarif.Validation/ResultLogBuilder.cs [hunk 12]
```diff
@@ -85,7 +85,10 @@ public class ResultLogBuilder: IDisposable
                 true,           // Produce verbose output.
                 new[] { instanceFilePath, schemaFilePath },
                 false,          // Do not compute target hash.
-                null);          // The version of this tool has no prerelease info.
+                null,          // The version of this tool has no prerelease info.
+                null);        // no invocation tokens to redact
```
Comment (lgolding) [inline src/Sarif.Validation/ResultLogBuilder.cs]: ... or you could use named parameters for documentation (`verbose: true`, _etc._)

File: src/Sarif/Converters/FxCopConverter.cs [hunk 13]
```diff
@@ -42,9 +42,12 @@ public void Convert(Stream input, IResultLogWriter output)
             }
 
             ToolInfo toolInfo = new ToolInfo();
-            RunInfo runInfo = new RunInfo();
             toolInfo.Name = "FxCop";
-            output.WriteToolAndRunInfo(toolInfo, runInfo);
+            output.WriteToolInfo(toolInfo);
+
+            // We can't infer/produce a runInfo object
```
Comment (lgolding) [inline src/Sarif/Converters/FxCopConverter.cs]: Why?

File: src/Sarif/IResultLogWriter.cs [hunk 14]
```diff
@@ -8,8 +8,24 @@ namespace Microsoft.CodeAnalysis.Sarif
...
+        /// Initialize the current output log.
+        /// </summary>
+        void Initialize();
+
+        /// <summary>Writes tool information to the log.</summary>
+        /// <exception cref="IOException">A file IO error occured. Clients implementing
+        /// <see cref="IToolFileConverter"/> should allow these exceptions to propagate.</exception>
+        /// <exception cref="InvalidOperationException">Thrown if the tool info block has already been
+        /// written.</exception>
+        /// <exception cref="ArgumentNullException">Thrown if <paramref name="info"/> is null.</exception>
```
Comment (lgolding) [inline src/Sarif/IResultLogWriter.cs]: `<paramref name="toolinfo" />`

File: src/Sarif/IResultLogWriter.cs [hunk 15]
```diff
@@ -8,8 +8,24 @@ namespace Microsoft.CodeAnalysis.Sarif
...
+        /// <param name="toolInfo">The tool information to write.</param>
+        /// <param name="runInfo">The run information to write.</param>
```
Comment (lgolding) [inline src/Sarif/IResultLogWriter.cs]: `runInfo` is not a parameter of this method.

File: src/Sarif/IResultLogWriter.cs [hunk 16]
```diff
@@ -17,7 +33,15 @@ public interface IResultLogWriter
         /// <exception cref="ArgumentNullException">Thrown if <paramref name="info"/> is null.</exception>
         /// <param name="toolInfo">The tool information to write.</param>
         /// <param name="runInfo">The run information to write.</param>
-        void WriteToolAndRunInfo(ToolInfo toolInfo, RunInfo runInfo);
+        void WriteRunInfo(RunInfo runInfo);
+
```
Comment (lgolding) [inline src/Sarif/IResultLogWriter.cs]: `<paramref name="runInfo" /> on Line 33.

`toolInfo` is not a parameter of this method.

File: src/Sarif/IResultLogWriter.cs [hunk 17]
```diff
@@ -17,7 +33,15 @@ public interface IResultLogWriter
...
+        /// <summary>
+        /// Initialize the results array associated with the current output log. SARIF producers that
+        /// are explicitly generating results (as opposed to other SARIF scenarios such as publishing
+        /// rules metadata) should proactively call this method in order to ensure that an explicit 
+        /// (but empty) results array exists in the log when no literal results were produced.
+        /// </summary>
+        void OpenResults();
```
Comment (lgolding) [inline src/Sarif/IResultLogWriter.cs]: I'm a little concerned about this API. For one thing, SDK users might wonder why there's no `CloseResults` method. How about `CreateEmptyResultsArray`?

File: src/Sarif/Readers/DateTimeConverter.cs [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+        }
+
+        public override bool CanConvert(Type objectType)
+        {
+            return objectType == typeof(DateTime);
+        }
+
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+        {
+            return DateTime.Parse((string)reader.Value);
```
Comment (lgolding) [inline src/Sarif/Readers/DateTimeConverter.cs]: Use the overload with `CultureInfo.InvariantCulture`.

File: src/Sarif/SarifResources.resx [hunk 19]
```diff
@@ -201,4 +198,16 @@
...
     <value>A CppCheck entry must have a nonzero number of locations.</value>
   </data>
+  <data name="InvalidState" xml:space="preserve">
+    <value>One or more invalid states were detected during serialization: {0}</value>
+  </data>
+  <data name="ResultsSerializationNotComplete" xml:space="preserve">
+    <value>Object cannot be serialized until results serialization is completed.</value>
+  </data>
+  <data name="RunInfoAlreadyWritten" xml:space="preserve">
+    <value>RunInfo cannot be written at this location (it has already been written).</value>
```
Comment (lgolding) [inline src/Sarif/SarifResources.resx]: "`RunInfo has already been written. It cannot be written again.`"

File: src/Sarif/SarifResources.resx [hunk 20]
```diff
@@ -201,4 +198,16 @@
...
+  </data>
+  <data name="ToolInfoAlreadyWritten" xml:space="preserve">
+    <value>ToolInfo cannot be written at this location (it has already been written).</value>
```
Comment (lgolding) [inline src/Sarif/SarifResources.resx]: ... and similarly for `ToolInfo`.

File: src/Sarif/SarifResources.resx [hunk 21]
```diff
@@ -201,4 +198,16 @@
   <data name="CppCheckMissingLocation" xml:space="preserve">
     <value>A CppCheck entry must have a nonzero number of locations.</value>
   </data>
+  <data name="InvalidState" xml:space="preserve">
+    <value>One or more invalid states were detected during serialization: {0}</value>
```
Comment (lgolding) [inline src/Sarif/SarifResources.resx]: I doubt an end user will understand this message. Is it saying that we tried to read a file whose contents were invalid? Is it saying that we called methods of the logging API in the wrong order? Something else? Whichever it is, let's say that.

File: src/Sarif/Writers/ResultLogJsonWriter.cs [hunk 22]
```diff
@@ -37,6 +44,22 @@ public ResultLogJsonWriter(JsonWriter jsonWriter)
             _serializer.ContractResolver = SarifContractResolver.Instance;
         }
 
+        public void Initialize()
+        {
+            this.EnsureStateNotAlreadySet(Conditions.Disposed | Conditions.Initialized);
+
+            _jsonWriter.WriteStartObject(); // Begin: resultLog
```
Comment (lgolding) [inline src/Sarif/Writers/ResultLogJsonWriter.cs]: `sarifLog`

File: src/Sarif/Writers/ResultLogJsonWriter.cs [hunk 23]
```diff
@@ -124,38 +161,50 @@ public void WriteResult(Result result)
...
             {
-                throw new InvalidOperationException(SarifResources.CannotWriteResultToolInfoMissing);
+                OpenResults();
             }
 
-            Debug.Assert(_writeState == State.WritingResults);
             _serializer.Serialize(_jsonWriter, result, typeof(Result));
         }
 
         public void CloseResults()
```
Comment (lgolding) [inline src/Sarif/Writers/ResultLogJsonWriter.cs]: So `ResultLogJsonWriter` has a public `CloseResults` method, but it's not part of the `IResultLogWriter` interface? I don't understand the reason for that design choice.

File: src/Sarif/Writers/ResultLogJsonWriter.cs [hunk 24]
```diff
@@ -167,14 +216,36 @@ public void Dispose()
...
+            }
+        }
+
+        private void EnsureStateNotAlreadySet(Conditions invalidConditions)
+        {
+            Conditions observedInvalidConditions = _writeConditions & invalidConditions;
+            if (observedInvalidConditions != Conditions.None)
+            {
+                // 	InvalidState	One or more invalid states were detected during serialization: {0}	
+                throw new InvalidOperationException(string.Format(SarifResources.InvalidState, observedInvalidConditions));
```
Comment (lgolding) [inline src/Sarif/Writers/ResultLogJsonWriter.cs]: Ok, now I know what it means. I think we need to be much friendlier. (I'll come back to this comment in a minute...)

Comment (lgolding) [inline src/Sarif/Writers/ResultLogJsonWriter.cs]: Ok, I'm back. I see that you have some friendly error message resource strings like `RunInfoAlreadyWritten`, but you never use them. And with the existing design, where this general-purpose method `EnsureStateNotAlreadySet` reports the error, it's hard to use the friendly special-purpose error messages. This is an interesting case where our life is harder because we have a single method with two reponsibilities: detect the error, and report the error.

I would replace this method with the following:

```
private bool StateIsAlreadySet(Conditions existingConditions)
{
    return (_writeConditions & existingConditions) != Conditions.None;
}
```

Then in each of the places where you're currently calling `EnsureStateNotAlreadySet`, I'd report the appropriate friendly message. For example, in `WriteToolInfo`:

```
    // This block could be put in a helper method EnsureNotDisposed().
    if (StateIsAlreadySet(Conditions.Disposed))
    {
        throw new ObjectDisposedException("The log writer object cannot be used because it has already been disposed.");
    }

    if (StateIsAlreadySet(Conditions.ToolInfoWritten))
    {
         throw new InvalidOperationException("You cannot write ToolInfo to the log because it has already been written.");
    }
```

Comment (lgolding) [inline src/Sarif/Writers/ResultLogJsonWriter.cs]: Now I also understand why you changed the expected exceptions in some of your tests. You had this general purpose method that _always throws the same exception_. So you didn't get the exception you really wanted. With my proposed design, you get `ObjectDisposedException` when you should, and `InvalidOperationException` when you should.

File: src/Sarif.Driver/Sdk/AnalyzeCommandBase.cs (hunk 1 above)
Comment (michaelcfanning) [inline src/Sarif.Driver/Sdk/AnalyzeCommandBase.cs]: yes will fix

File: src/Sarif.Driver/Sdk/SarifLogger.cs (hunk 2 above)
Comment (michaelcfanning) [inline src/Sarif.Driver/Sdk/SarifLogger.cs]: s/be public. fixed.

File: src/Sarif.Driver/Sdk/SarifLogger.cs (hunk 3 above)
Comment (michaelcfanning) [inline src/Sarif.Driver/Sdk/SarifLogger.cs]: filed #153

File: src/Sarif.Driver/Sdk/SarifLogger.cs (hunk 4 above)
Comment (michaelcfanning) [inline src/Sarif.Driver/Sdk/SarifLogger.cs]: good catch

File: src/Sarif/Writers/ResultLogJsonWriter.cs (hunk 6 above)
Comment (michaelcfanning) [inline src/Sarif/Writers/ResultLogJsonWriter.cs]: No, we allow some resilience for writers that stream individual results without trying to manage the results array. I added a comment.

File: src/Sarif.UnitTests/Writers/ResultLogJsonWriterTests.cs (hunk 8 above)
Comment (michaelcfanning) [inline src/Sarif.UnitTests/Writers/ResultLogJsonWriterTests.cs]: fixed

File: src/Sarif.UnitTests/Writers/ResultLogJsonWriterTests.cs (hunk 9 above)
Comment (michaelcfanning) [inline src/Sarif.UnitTests/Writers/ResultLogJsonWriterTests.cs]: Because we now filter all state checks (including asking the question 'are you already disposed?' through a common helper. rather than over-complicating the exception raising signature, I simply reduce all state confusion to InvalidOperationException

Comment (lgolding) [inline src/Sarif.UnitTests/Writers/ResultLogJsonWriterTests.cs]: Yup, I figured that out later on. See my proposal below.

File: src/Sarif/Converters/FxCopConverter.cs (hunk 13 above)
Comment (michaelcfanning) [inline src/Sarif/Converters/FxCopConverter.cs]: writing of toolInfo and runInfo are now broken into distinct steps. also, previously, writing the toolInfo initialized the log file (by writing root runLog and version details). as of your recently merged change, we do, in fact, produce a runInfo object in some circumstances (1 or more results generated)

File: src/Sarif/IResultLogWriter.cs (hunk 14 above)
Comment (michaelcfanning) [inline src/Sarif/IResultLogWriter.cs]: yes fixed

File: src/Sarif/IResultLogWriter.cs (hunk 16 above)
Comment (michaelcfanning) [inline src/Sarif/IResultLogWriter.cs]: someone should run all the roslyn analyzers against this code :P

File: src/Sarif/IResultLogWriter.cs (hunk 17 above)
Comment (michaelcfanning) [inline src/Sarif/IResultLogWriter.cs]: There is a CloseResults api, just neglected to add it to the interface. Now fixed.

File: src/Sarif/Readers/DateTimeConverter.cs (hunk 18 above)
Comment (michaelcfanning) [inline src/Sarif/Readers/DateTimeConverter.cs]: Oops. Yes. Also in the ToString call for WriteJson. You just fixed someone's nightmarish bug where a log file was produced and then subsequently consumed in a different culture but a time window that would make everything go haywire.

File: src/Sarif/SarifResources.resx (hunk 20 above)
Comment (michaelcfanning) [inline src/Sarif/SarifResources.resx]: yes

File: src/Sarif/Writers/ResultLogJsonWriter.cs (hunk 22 above)
Comment (michaelcfanning) [inline src/Sarif/Writers/ResultLogJsonWriter.cs]: yes

File: src/Sarif/Writers/ResultLogJsonWriter.cs (hunk 23 above)
Comment (michaelcfanning) [inline src/Sarif/Writers/ResultLogJsonWriter.cs]: fixed

File: src/Sarif/Writers/ResultLogJsonWriter.cs (hunk 24 above)
Comment (michaelcfanning) [inline src/Sarif/Writers/ResultLogJsonWriter.cs]: point noted on useful new messages, we can file a bug.

File: src/Sarif.Driver.UnitTests/Sdk/FormatForVisualStudioTests.cs [hunk 25]
```diff
@@ -15,7 +15,6 @@ public class FormatForVisualStudioTests
         private const string TestRuleId = "TST0001";
         private const string TestFormatSpecifier = "testFormatSpecifier";
         private const string TestAnalysisTarget = @"C:\dir\file";
-        private static readonly string DisplayedTarget = TestAnalysisTarget.Replace('\\', '/');
```
Comment (michaelcfanning) [inline src/Sarif.Driver.UnitTests/Sdk/FormatForVisualStudioTests.cs]: @lgolding, note I fixed the forward slashes that appeared in the VS output (which was certainly not by design).
